huggingface_auth.py

```diff
@@ -0,0 +1,132 @@
+from hivemind.utils.crypto import RSAPublicKey
+from hivemind.utils.logging import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class HuggingFaceAuthorizer(TokenAuthorizerBase):
+    _AUTH_SERVER_URL = 'https://collaborative-training-auth.huggingface.co'
+
+    def __init__(self, experiment_id: int, username: str, password: str):
+        super().__init__()
+
+        self._experiment_id = experiment_id
+        self._username = username
+        self._password = password
+
+        self._authority_public_key = None
+        self._hf_api = HfApi()
+
+    async def get_token(self) -> AccessToken:
+        token = self._hf_api.login(self._username, self._password)
+
+        try:
+            url = f'{self._AUTH_SERVER_URL}/api/experiments/join/{self._experiment_id}/'
+            headers = {'Authorization': f'Bearer {token}'}
+            response = requests.put(url, headers=headers, json={
+                'experiment_join_input': {
+                    'peer_public_key': self.local_public_key.to_bytes().decode(),
+                },
+            })
+
+            response.raise_for_status()
+            response = response.json()
+
+            self._authority_public_key = RSAPublicKey.from_bytes(response['auth_server_public_key'].encode())
+
+            token_dict = response['hivemind_access']
+            access_token = AccessToken()
+            access_token.username = token_dict['username']
+            access_token.public_key = token_dict['peer_public_key'].encode()
+            access_token.expiration_time = str(datetime.fromisoformat(token_dict['expiration_time']))
+            access_token.signature = token_dict['signature'].encode()
+
+            logger.info(f'Access for user {access_token.username} '
+                        f'has been granted until {access_token.expiration_time} UTC')
+            return access_token
+        finally:
+            self._hf_api.logout(token)
+
+    def add_collaborator(self) -> None:
+        # This is a temporary workaround necessary until the experiment invite tokens are implemented.
+        # It is not intended to be secure and designed to test the authorization code
+        # without complicating the new user's joining procedure.
+
+        token = self._hf_api.login('robot-bengali', base64.b64decode(b'aGdKQlViTDMzd2h2').decode())
+
+        try:
+            url = f'{self._AUTH_SERVER_URL}/api/experiments/{self._experiment_id}/'
+            headers = {'Authorization': f'Bearer {token}'}
+            response = requests.put(url, headers=headers, json={
+                'experiment_full_update': {
+                    'added_collaborators': [{'username': self._username}],
+                },
+            })
+
+            response.raise_for_status()
+            logger.info(f'User {self._username} has been added to collaborators of '
+                        f'the experiment {self._experiment_id}')
+        finally:
+            self._hf_api.logout(token)
```

Comment:
Just to remember and share with you all my vision, it seems to me that this method would not be useful anymore for the version with whitelisting. :)

Reply:
Sure :)